dotnet/src/Extensions/TemplateEngine.Basic/BasicPromptTemplate.cs

```diff
@@ -0,0 +1,167 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using Microsoft.SemanticKernel.Diagnostics;
+using Microsoft.SemanticKernel.Orchestration;
+using Microsoft.SemanticKernel.TemplateEngine.Basic.Blocks;
+
+namespace Microsoft.SemanticKernel.TemplateEngine.Basic;
+
+/// <summary>
+/// Given a prompt, that might contain references to variables and functions:
+/// - Get the list of references
+/// - Resolve each reference
+///   - Variable references are resolved using the context variables
+///   - Function references are resolved invoking those functions
+///     - Functions can be invoked passing in variables
+///     - Functions do not receive the context variables, unless specified using a special variable
+///     - Functions can be invoked in order and in parallel so the context variables must be immutable when invoked within the template
+/// </summary>
+public sealed class BasicPromptTemplate : IPromptTemplate
+{
+    /// <summary>
+    /// Constructor for PromptTemplate.
+    /// </summary>
+    /// <param name="templateString">Prompt template string.</param>
+    /// <param name="promptTemplateConfig">Prompt template configuration</param>
+    /// <param name="loggerFactory">Logger factory</param>
+    public BasicPromptTemplate(string templateString, PromptTemplateConfig promptTemplateConfig, ILoggerFactory? loggerFactory = null)
```

Comment:
In what situation does someone create one of these directly? Could this whole type be internal instead?

Reply:
If I'm using the agregator to combine support for the basic (default) format plus Handlebars, plus Fluid I would do something like this:

```
var target = new AggregatorPromptTemplateFactory(
    new BasicPromptTemplateFactory(), 
    new HandlebarsPromptTemplateFactory(), 
    new FluidPromptTemplateFactory());
```